feat: support for opening/closing private broker channels

msgmaxim opened the pull request:
# Pull Request

Closes: PRO-1748

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have written sufficient tests.

## Summary

- Added a storage map `AccountId` -> `ChannelId`
- Added Extrinsics to open/close private channels + tests + benchmarks

Out of scope (these seem like may need some further discussion):
- Charging fee/rent for opening a private channel/keeping it open
- Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

May need to rebase this after #5357 is merged.

dandanlen commented on state-chain/pallets/cf-ingress-egress/src/lib.rs:
```diff
@@ -582,6 +582,10 @@ pub mod pallet {
 
+	#[pallet::storage]
+	pub(crate) type BrokerPrivateChannels<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Twox64Concat, T::AccountId, ChannelId, OptionQuery>;
```
This can use the Identity hasher.

```suggestion
		StorageMap<_, Identity, T::AccountId, ChannelId, OptionQuery>;
```

kylezs commented on state-chain/pallets/cf-ingress-egress/src/lib.rs:
```diff
@@ -1310,6 +1326,50 @@ pub mod pallet {
+			);
+
+			// TODO: burn fee for opening a channel?
+			let next_channel_id =
+				ChannelIdCounter::<T, I>::try_mutate::<_, Error<T, I>, _>(|id| {
+					*id = id.checked_add(1).ok_or(Error::<T, I>::ChannelIdsExhausted)?;
+					Ok(*id)
+				})?;
```
we should make this a function, is duplicate

msgmaxim replied:
Deduplicated

kylezs commented on state-chain/pallets/cf-ingress-egress/src/lib.rs:
```diff
@@ -1310,6 +1326,50 @@ pub mod pallet {
+			let broker_id = T::AccountRoleRegistry::ensure_broker(origin)?;
+
+			ensure!(
+				!BrokerPrivateChannels::<T, I>::contains_key(&broker_id),
```
Should we also restrict this to be just for Bitcoin - it's not useful for any other chains right? probably more importantly is that not restricting it to Bitcoin would be confusing when reading the code, since it's an instantiable pallet, we'd assume it's for all instances.

msgmaxim replied:
Added the check.

kylezs commented on state-chain/pallets/cf-ingress-egress/src/lib.rs:
```diff
@@ -1310,6 +1326,50 @@ pub mod pallet {
+				Error::<T, I>::PrivateChannelExistsForBroker
+			);
+
+			// TODO: burn fee for opening a channel?
```
I see this burning/rent is mentioned in the ticket, but also that this PR closes that ticket? so maybe we create another to address this separately.

msgmaxim replied:
I will create tickets once we decide what should be in the scope of this PR and what isn't (I mentioned this in the description).

kylezs replied:
> Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

Ah ok, that seemed to suggest was just about the account deletion

dandanlen commented:
> Closing private channels upon account deletion (maybe this isn't necessary for an MVP?)
(will create separate linear tickets for these if we agree to leave them out of this PR)

I feel like we can just add this here. We need it and it's not a lot of work to add. Not sure what we gain by waiting. 

(Sorry was sure I left this comment yesterday...)

dandanlen commented:
Forgot to mention - I think charging/not charging a fee needs a little more discussion, happy to leave this out for now.

msgmaxim commented:
Addressed all comments now. Ready for re-review @dandanlen @kylezs

kylezs commented on state-chain/pallets/cf-swapping/src/lib.rs:
```diff
@@ -1054,6 +1075,34 @@ pub mod pallet {
+		pub fn open_private_channel(origin: OriginFor<T>, chain: ForeignChain) -> DispatchResult {
+			let broker_id = T::AccountRoleRegistry::ensure_broker(origin)?;
+
+			ensure!(chain == ForeignChain::Bitcoin, Error::<T>::NoPrivateChannelsForChain);
```
Now we only inject the BitcoinIngressEgress as the PrivateChannelManager, so it's only possible to create bitcoin channel, or inversely, if we injected the wrong instance of the ingress egress then we would put bitcoin in here, and then open an Ethereum channel. 
 So perhaps would be better to have this be:
```fn open_private_btc_channel(origin)``` and we can remove this ensure, same below.

dandanlen replied:
Was going to suggest the same thing - the only benefit of the `chain` parameter is in case we want to add another utxo chain... but that seems unlikely right now and in the meantime it's not used for anything.

kylezs commented on state-chain/pallets/cf-swapping/src/tests.rs:
```diff
@@ -1618,3 +1652,64 @@ mod swap_batching {
+	#[test]
+	fn close_private_channel() {
+		new_test_ext().execute_with(|| {
+			// Only brokers can  channels
```
```suggestion
			// Only brokers can open channels
```

kylezs reviewed (approved):
Nice tests 👌  - would let Dan approve before merge

dandanlen commented on state-chain/pallets/cf-swapping/src/lib.rs:
```diff
@@ -1054,6 +1075,34 @@ pub mod pallet {
+		pub fn close_private_channel(origin: OriginFor<T>, chain: ForeignChain) -> DispatchResult {
+			let broker_id = T::AccountRoleRegistry::ensure_broker(origin)?;
+
+			ensure!(chain == ForeignChain::Bitcoin, Error::<T>::NoPrivateChannelsForChain);
```
Same here, including the `chain` param seems unnecessary.

dandanlen reviewed (commented):
I think we might as well make it btc-specific and remove the chain params to make it clearer.

msgmaxim commented:
Renamed the extrinsics in cf-swapping and removed the chain parameter. Note that I chose not to rename the methods inside cf-ingress-egress as I thought it would look odd (it would be misleading for non-btc instances to have a btc-specific method), but I added a comment to say that this functionality is expected to be used by bitcoin only. Importantly, we don't have a chain parameter and we made it more explicit to external users that this is bitcoin only feature. @dandanlen

kylezs commented:
> Note that I chose not to rename the methods inside cf-ingress-egress

I think that makes sense 👍

dandanlen reviewed (commented):
LGTM - only thing is that looking at the code now, i'm not sure why we put the storage in the ingess-egress pallet. All it needs to do is return an incremented channel id. We could store it in the swapping pallet and then we have a smaller internal API surface (no need for close/lookup methods in the trait). 

Will leave it up to you @msgmaxim whether to make this change or not. I think it would reduce complexity but having the 'channel' (even though it's just the id) in the same place as other channels also kind of makes sense.

msgmaxim commented:
> LGTM - only thing is that looking at the code now, i'm not sure why we put the storage in the ingess-egress pallet. All it needs to do is return an incremented channel id. We could store it in the swapping pallet and then we have a smaller internal API surface (no need for close/lookup methods in the trait).
> 
> Will leave it up to you @msgmaxim whether to make this change or not. I think it would reduce complexity but having the 'channel' (even though it's just the id) in the same place as other channels also kind of makes sense.

Made these changes now @dandanlen
